File: packages/logic/src/mechanics/talismanCosts.ts
// Talisman fragment-cost progression formulas, lifted from
// packages/web_ui/src/Talismans.ts. Both formulas are pure functions over a
// Decimal `baseMult` and an integer `level` — no player state, no globals.
// They return the fragment cost map for the NEXT level of the talisman.
//
// `regularCostProgression` is the cubic-tier formula used by most talismans:
// each fragment tier starts contributing at a level threshold (shard at 0,
// common at 30, uncommon at 60, etc.) and the base multiplier itself ramps
// up at higher levels (120/150/180). `exponentialCostProgression` is the
// alternative used by a handful of talismans whose cost grows as `ratio^level`
// instead of `level^3` — same tier-threshold scheme, different growth curve.

import { Decimal } from '../math/bignum'

export type TalismanCraftItems =
  | 'shard'
  | 'commonFragment'
  | 'uncommonFragment'
  | 'rareFragment'
  | 'epicFragment'
  | 'legendaryFragment'
  | 'mythicalFragment'

export type TalismanCraftCosts = Record<TalismanCraftItems, Decimal>

/**
 * Cubic-tier cost progression. For each tier, the cost is
 * `floor((level - threshold)^3 / divisor + 1) * priceMult`, clamped at zero
 * below the tier threshold. `priceMult` itself grows piecewise past levels
 * 120 / 150 / 180. Returns a 0-cost map until each tier unlocks (shard at 0,
 * common at 30, uncommon at 60, rare at 90, epic at 120, legendary/mythical
 * at 150).
 */
export function regularCostProgression (baseMult: Decimal, level: number): TalismanCraftCosts {
  let priceMult = baseMult
  if (level >= 120) {
    priceMult = priceMult.times((level - 90) / 30)
  }
  if (level >= 150) {
    priceMult = priceMult.times((level - 120) / 30)
  }
  if (level >= 180) {
    priceMult = priceMult.times((level - 170) / 10)
  }

  const shardCost = Decimal.pow(level, 3).times(1 / 8).plus(1).floor().times(priceMult)
  const commonCost = level >= 30
    ? Decimal.pow(level - 30, 3).times(1 / 32).plus(1).floor().times(priceMult)
    : new Decimal(0)
  const uncommonCost = level >= 60
    ? Decimal.pow(level - 60, 3).times(1 / 384).plus(1).floor().times(priceMult)
    : new Decimal(0)
  const rareCost = level >= 90
    ? Decimal.pow(level - 90, 3).times(1 / 500).plus(1).floor().times(priceMult)
    : new Decimal(0)
  const epicCost = level >= 120
    ? Decimal.pow(level - 120, 3).times(1 / 375).plus(1).floor().times(priceMult)
    : new Decimal(0)
  const legendaryCost = level >= 150
    ? Decimal.pow(level - 150, 3).times(1 / 192).plus(1).floor().times(priceMult)
    : new Decimal(0)
  const mythicalCost = level >= 150
    ? Decimal.pow(level - 150, 3).times(1 / 1280).plus(1).floor().times(priceMult)
    : new Decimal(0)

  return {
    shard: Decimal.max(0, shardCost),
    commonFragment: Decimal.max(0, commonCost),
    uncommonFragment: Decimal.max(0, uncommonCost),
    rareFragment: Decimal.max(0, rareCost),
    epicFragment: Decimal.max(0, epicCost),
    legendaryFragment: Decimal.max(0, legendaryCost),
    mythicalFragment: Decimal.max(0, mythicalCost)
  }
}

/**
 * Exponential cost progression. For each tier the cost is
 * `floor(ratio^(level - threshold) * baseMult * tierConstant)`. The tier
 * constants are fixed (100 / 50 / 25 / 20 / 15 / 10 / 5) and the tier
 * thresholds match `regularCostProgression`. `ratio` is supplied by the
 * caller — common values are 2, 10, 1e5, 1e8 (each used by one talisman).
 */
export function exponentialCostProgression (
  baseMult: Decimal,
  level: number,
  ratio: number
): TalismanCraftCosts {
  const baseMultDecimal = new Decimal(baseMult)

  return {
    shard: Decimal.pow(ratio, level).times(baseMultDecimal).times(100).floor(),
    commonFragment: level >= 30
      ? Decimal.pow(ratio, level - 30).times(baseMultDecimal).times(50).floor()
      : new Decimal(0),
    uncommonFragment: level >= 60
      ? Decimal.pow(ratio, level - 60).times(baseMultDecimal).times(25).floor()
      : new Decimal(0),
    rareFragment: level >= 90
      ? Decimal.pow(ratio, level - 90).times(baseMultDecimal).times(20).floor()
      : new Decimal(0),
    epicFragment: level >= 120
      ? Decimal.pow(ratio, level - 120).times(baseMultDecimal).times(15).floor()
      : new Decimal(0),
    legendaryFragment: level >= 150
      ? Decimal.pow(ratio, level - 150).times(baseMultDecimal).times(10).floor()
      : new Decimal(0),
    mythicalFragment: level >= 150
      ? Decimal.pow(ratio, level - 150).times(baseMultDecimal).times(5).floor()
      : new Decimal(0)
  }
}

File: crates/synergismforkd_logic/src/mechanics/talisman_costs.rs
//! Talisman fragment-cost progression formulas.
//!
//! Verbatim port of
//! `legacy_core_split/packages/logic/src/mechanics/talismanCosts.ts`.
//! Both formulas are pure functions over a `Decimal` `base_mult` and
//! an integer `level` — no player state, no globals. They return the
//! fragment cost map for the **next** level of the talisman.
//!
//! [`regular_cost_progression`] is the cubic-tier formula used by most
//! talismans: each fragment tier starts contributing at a level
//! threshold (shard at 0, common at 30, uncommon at 60, etc.) and the
//! base multiplier itself ramps up at higher levels (120/150/180).
//! [`exponential_cost_progression`] is the alternative used by a
//! handful of talismans whose cost grows as `ratio^level` instead of
//! `level^3` — same tier-threshold scheme, different growth curve.

use synergismforkd_bignum::Decimal;

/// Fragment cost map for the next talisman level. Each field is the
/// cost of one fragment kind.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct TalismanCraftCosts {
    /// Cost in talisman shards.
    pub shard: Decimal,
    /// Cost in common fragments.
    pub common_fragment: Decimal,
    /// Cost in uncommon fragments.
    pub uncommon_fragment: Decimal,
    /// Cost in rare fragments.
    pub rare_fragment: Decimal,
    /// Cost in epic fragments.
    pub epic_fragment: Decimal,
    /// Cost in legendary fragments.
    pub legendary_fragment: Decimal,
    /// Cost in mythical fragments.
    pub mythical_fragment: Decimal,
}

fn cubic_tier_cost(level: f64, threshold: f64, divisor: f64, price_mult: Decimal) -> Decimal {
    if level < threshold {
        return Decimal::zero();
    }
    let cost = (Decimal::from_finite(level - threshold).pow(Decimal::from_finite(3.0))
        / Decimal::from_finite(divisor)
        + Decimal::one())
    .floor()
        * price_mult;
    cost.max(Decimal::zero())
}

/// Cubic-tier cost progression. For each tier, the cost is
/// `floor((level - threshold)^3 / divisor + 1) * price_mult`, clamped
/// at zero below the tier threshold. `price_mult` itself grows
/// piecewise past levels `120 / 150 / 180`. Returns a 0-cost map until
/// each tier unlocks (shard at 0, common at 30, uncommon at 60, rare
/// at 90, epic at 120, legendary/mythical at 150).
#[must_use]
pub fn regular_cost_progression(base_mult: Decimal, level: f64) -> TalismanCraftCosts {
    let mut price_mult = base_mult;
    if level >= 120.0 {
        price_mult *= Decimal::from_finite((level - 90.0) / 30.0);
    }
    if level >= 150.0 {
        price_mult *= Decimal::from_finite((level - 120.0) / 30.0);
    }
    if level >= 180.0 {
        price_mult *= Decimal::from_finite((level - 170.0) / 10.0);
    }

    // Shard has no threshold gate — uses raw `level` (not level - 0).
    let shard_cost = (Decimal::from_finite(level).pow(Decimal::from_finite(3.0))
        / Decimal::from_finite(8.0)
        + Decimal::one())
    .floor()
        * price_mult;

    TalismanCraftCosts {
        shard: shard_cost.max(Decimal::zero()),
        common_fragment: cubic_tier_cost(level, 30.0, 32.0, price_mult),
        uncommon_fragment: cubic_tier_cost(level, 60.0, 384.0, price_mult),
        rare_fragment: cubic_tier_cost(level, 90.0, 500.0, price_mult),
        epic_fragment: cubic_tier_cost(level, 120.0, 375.0, price_mult),
        legendary_fragment: cubic_tier_cost(level, 150.0, 192.0, price_mult),
        mythical_fragment: cubic_tier_cost(level, 150.0, 1_280.0, price_mult),
    }
}

fn exponential_tier_cost(
    level: f64,
    threshold: f64,
    ratio: f64,
    base_mult: Decimal,
    tier_constant: f64,
) -> Decimal {
    if level < threshold {
        return Decimal::zero();
    }
    (Decimal::from_finite(ratio).pow(Decimal::from_finite(level - threshold))
        * base_mult
        * Decimal::from_finite(tier_constant))
    .floor()
}

/// Exponential cost progression. For each tier the cost is
/// `floor(ratio^(level - threshold) * base_mult * tier_constant)`. The
/// tier constants are fixed (`100 / 50 / 25 / 20 / 15 / 10 / 5`) and
/// the tier thresholds match [`regular_cost_progression`]. `ratio` is
/// supplied by the caller — common values are `2, 10, 1e5, 1e8` (each
/// used by one talisman).
#[must_use]
pub fn exponential_cost_progression(
    base_mult: Decimal,
    level: f64,
    ratio: f64,
) -> TalismanCraftCosts {
    TalismanCraftCosts {
        // Shard has no threshold gate — uses raw `level`.
        shard: (Decimal::from_finite(ratio).pow(Decimal::from_finite(level))
            * base_mult
            * Decimal::from_finite(100.0))
        .floor(),
        common_fragment: exponential_tier_cost(level, 30.0, ratio, base_mult, 50.0),
        uncommon_fragment: exponential_tier_cost(level, 60.0, ratio, base_mult, 25.0),
        rare_fragment: exponential_tier_cost(level, 90.0, ratio, base_mult, 20.0),
        epic_fragment: exponential_tier_cost(level, 120.0, ratio, base_mult, 15.0),
        legendary_fragment: exponential_tier_cost(level, 150.0, ratio, base_mult, 10.0),
        mythical_fragment: exponential_tier_cost(level, 150.0, ratio, base_mult, 5.0),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn regular_progression_locks_tiers_below_threshold() {
        let costs = regular_cost_progression(Decimal::one(), 29.0);
        assert!(costs.shard > Decimal::zero());
        assert_eq!(costs.common_fragment, Decimal::zero());
        assert_eq!(costs.uncommon_fragment, Decimal::zero());
        assert_eq!(costs.rare_fragment, Decimal::zero());
        assert_eq!(costs.epic_fragment, Decimal::zero());
        assert_eq!(costs.legendary_fragment, Decimal::zero());
        assert_eq!(costs.mythical_fragment, Decimal::zero());
    }

    #[test]
    fn regular_progression_common_unlocks_at_30() {
        let below = regular_cost_progression(Decimal::one(), 29.0);
        let at = regular_cost_progression(Decimal::one(), 30.0);
        assert_eq!(below.common_fragment, Decimal::zero());
        // Level 30: (0)^3 / 32 + 1 = 1; floor(1) = 1.
        assert_eq!(at.common_fragment.to_number(), 1.0);
    }

    #[test]
    fn regular_progression_price_mult_ramps_at_120() {
        // base_mult = 1; at level 120, multiplier becomes (120-90)/30 = 1
        // (still 1, but the ramp factor IS applied — proves the branch fires).
        // To test the ramp visibly: at 150, multiplier = (150-90)/30 * (150-120)/30
        // = 2 * 1 = 2.
        let at_150 = regular_cost_progression(Decimal::one(), 150.0);
        let below_120 = regular_cost_progression(Decimal::one(), 119.0);
        // At level 120 the multiplier kicks in; at 150 it's even bigger.
        // We just check that level-150 shard cost is meaningfully above the
        // level-119 shard cost (which it would be even without the ramp).
        assert!(at_150.shard > below_120.shard);
    }

    #[test]
    fn exponential_progression_locks_tiers_below_threshold() {
        let costs = exponential_cost_progression(Decimal::one(), 29.0, 2.0);
        assert!(costs.shard > Decimal::zero());
        assert_eq!(costs.common_fragment, Decimal::zero());
    }

    #[test]
    fn exponential_progression_shard_scales_with_ratio() {
        // ratio = 2, level = 5, base = 1 → shard = floor(2^5 * 1 * 100) = 3200
        let costs = exponential_cost_progression(Decimal::one(), 5.0, 2.0);
        assert_eq!(costs.shard.to_number(), 3_200.0);
    }

    #[test]
    fn exponential_progression_common_unlocks_at_30() {
        let at = exponential_cost_progression(Decimal::one(), 30.0, 2.0);
        // ratio = 2, level - threshold = 0, base = 1, tier_constant = 50
        // floor(2^0 * 1 * 50) = 50
        assert_eq!(at.common_fragment.to_number(), 50.0);
    }
}
